File: bindings/deno/src/safe_ml.ts
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/**
 * SafeML - Machine learning activation functions and utilities.
 *
 * Thin FFI wrapper: all computation delegates to libproven (Idris 2 + Zig).
 *
 * @module
 */

import { getLib, ProvenStatus, statusToError } from './ffi.ts';
import { err, ok, type Result } from './result.ts';

/**
 * Safe ML operations backed by formally verified Idris 2 code.
 */
export class SafeML {
  /**
   * Sigmoid activation function: 1 / (1 + exp(-x)).
   *
   * @param x - Input value.
   * @returns Sigmoid output.
   */
  static sigmoid(x: number): number {
    const symbols = getLib();
    return symbols.proven_ml_sigmoid(x);
  }

  /**
   * ReLU activation function: max(0, x).
   *
   * @param x - Input value.
   * @returns ReLU output.
   */
  static relu(x: number): number {
    const symbols = getLib();
    return symbols.proven_ml_relu(x);
  }

  /**
   * Leaky ReLU activation function: x > 0 ? x : alpha * x.
   *
   * @param x - Input value.
   * @param alpha - Leak coefficient (default 0.01).
   * @returns Leaky ReLU output.
   */
  static leakyRelu(x: number, alpha = 0.01): number {
    const symbols = getLib();
    return symbols.proven_ml_leaky_relu(x, alpha);
  }

  /**
   * Clamp value to range [min, max].
   *
   * @param x - Input value.
   * @param min - Minimum.
   * @param max - Maximum.
   * @returns Clamped value.
   */
  static clamp(x: number, min: number, max: number): number {
    const symbols = getLib();
    return symbols.proven_ml_clamp(x, min, max);
  }

  /**
   * Softmax normalization.
   *
   * @param input - Input values.
   * @returns Result containing normalized output or an error string.
   */
  static softmax(input: Float64Array): Result<Float64Array> {
    const symbols = getLib();
    const output = new Float64Array(input.length);
    const status = symbols.proven_ml_softmax(input, output, input.length);
    if (status !== ProvenStatus.OK) return err(statusToError(status));
    return ok(output);
  }
}

File: bindings/java/src/main/java/io/github/hyperpolymath/proven/SafeML.java
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
package io.github.hyperpolymath.proven;

/**
 * Safe machine-learning activation functions via libproven FFI.
 *
 * Provides sigmoid, ReLU, leaky ReLU, and clamping.
 * All logic delegates to the Idris 2 verified implementation through
 * the Zig C ABI. No activation-function logic is reimplemented in Java.
 */
public final class SafeML {

    private static final ProvenLibrary LIB = ProvenLibrary.INSTANCE;

    private SafeML() {}

    /**
     * Sigmoid activation function: 1 / (1 + exp(-x)).
     *
     * @param x the input value
     * @return the sigmoid output
     */
    public static double sigmoid(double x) {
        return LIB.proven_ml_sigmoid(x);
    }

    /**
     * ReLU activation function: max(0, x).
     *
     * @param x the input value
     * @return the ReLU output
     */
    public static double relu(double x) {
        return LIB.proven_ml_relu(x);
    }

    /**
     * Leaky ReLU activation function: x if x > 0, else alpha * x.
     *
     * @param x     the input value
     * @param alpha the leak coefficient for negative inputs
     * @return the leaky ReLU output
     */
    public static double leakyRelu(double x, double alpha) {
        return LIB.proven_ml_leaky_relu(x, alpha);
    }

    /**
     * Clamp a value to [min, max].
     *
     * @param x      the value to clamp
     * @param minVal the minimum bound
     * @param maxVal the maximum bound
     * @return the clamped value
     */
    public static double clamp(double x, double minVal, double maxVal) {
        return LIB.proven_ml_clamp(x, minVal, maxVal);
    }
}
